### PR TITLE
Fix issue 599

### PR DESCRIPTION
**Pull request by gungnir479**
Modified containsAtLeastElementsIn(). Now it figures out the actual order of the expected objects when it is not in order. It remembers the actual order, stores the actual objects into a list. If it is not in order, retain the excepted list to get the actual order of the expected objects.

### CODE REVIEW

File: core/src/test/java/com/google/common/truth/IterableSubjectTest.java [hunk 1]
```diff
@@ -388,8 +388,10 @@ public void iterableContainsAtLeastInOrderWithFailure() {
     assertFailureKeys(
         "required elements were all found, but order was wrong",
         "expected order for required elements",
+        "actual order for required elements",
```
Comment (chaoren) [inline core/src/test/java/com/google/common/truth/IterableSubjectTest.java:391]: The error message after this change would be:

```
required elements were all found, but order was wrong
expected order for required elements: [<expected order>]
actual order for required elements  : [<actual order>]
but was                             : [<full list>>]
```

I don't think that makes sense grammatically.

File: core/src/main/java/com/google/common/truth/IterableSubject.java [hunk 2]
```diff
@@ -306,16 +307,21 @@ public final Ordered containsAtLeastElementsIn(Iterable<?> expectedIterable) {
...
-          public void inOrder() {
-            failWithActual(
-                simpleFact("required elements were all found, but order was wrong"),
-                fact("expected order for required elements", expected));
-          }
-        };
+    if (ordered) {
+      return IN_ORDER;
+    } else {
+      order.retainAll(expected);
```
Comment (chaoren) [inline core/src/main/java/com/google/common/truth/IterableSubject.java:313]: This list would be unnecessarily created and filtered even if `.inOrder()` is not called.

File: core/src/main/java/com/google/common/truth/IterableSubject.java [hunk 3]
```diff
@@ -271,6 +271,7 @@ public final Ordered containsAtLeast(
   @CanIgnoreReturnValue
   public final Ordered containsAtLeastElementsIn(Iterable<?> expectedIterable) {
     List<?> actual = Lists.newLinkedList(this.actual);
+    List<?> order = Lists.newArrayList(actual);
```
Comment (chaoren) [inline core/src/main/java/com/google/common/truth/IterableSubject.java:274]: `order` isn't a very good name for a list, IMO.

**Comment by gungnir479** [comment]
> Please sign the CLA.



> Please sign the CLA.

I have signed it :）

File: core/src/main/java/com/google/common/truth/IterableSubject.java [hunk 4]
```diff
@@ -306,16 +307,21 @@ public final Ordered containsAtLeastElementsIn(Iterable<?> expectedIterable) {
...
-          }
-        };
+    if (ordered) {
+      return IN_ORDER;
+    } else {
+      order.retainAll(expected);
+      return new Ordered() {
+        @Override
+        public void inOrder() {
+          failWithActual(
```
Comment (chaoren) [inline core/src/main/java/com/google/common/truth/IterableSubject.java:317]: `failWithActual()` will still append a `but was` fact at the end with the full contents of `actual`. That would make the error message extremely confusing and grammatically incorrect.

**Review by chaoren — CHANGES_REQUESTED** [review]
`Fix issue 599` isn't a very descriptive title for the PR or commit.

File: core/src/main/java/com/google/common/truth/IterableSubject.java (hunk 4 above)
Comment (gungnir479) [reply core/src/main/java/com/google/common/truth/IterableSubject.java:317]: OK，I will replace `failWithActual()` with `failWithoutActual()`

File: core/src/main/java/com/google/common/truth/IterableSubject.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gungnir479) [reply core/src/main/java/com/google/common/truth/IterableSubject.java:274]: OK, I will replace `order` with ``actualOrder`.

I've tried to delay the construction of it, but there is a problem: When the `this.actual` is a one snap iterator, we cannot construct `actualOrder` from it. If we delay the construction, the `actual` is already been changed and also cannot construct `actualOrder` from it (see `iterableContainsAtLeastInOrderWithOneShotIterableWrongOrder()` in IterableSubjectTest). I'm wondering if there is a better solution...

**Comment by chaoren** [comment]
Sorry, but I'll take over fixing this.

**Pull request closed without merging**